File: imp/objective-c/src/fit/FrameworkTest.m
//
//  FrameworkTest.m
//  FIT
//
//  Created by Steven Newton on Thu Sep 18 2003.
//  Copyright (c) 2002 Cunningham & Cunningham, Inc.
//  Released under the terms of the GNU General Public License version 2 or later.
//

#import "FrameworkTest.h"

@interface TestFixture : Fixture {
    int sampleInt;
    NSNumber *sampleInteger;
    NSNumber *sampleFloat;
    BOOL sampleBoolean;
    unichar ch;
    NSString* name;
}
- (NSString *) name;
- (void) name: (NSString *) s;
- (int) sampleInt;
- (void) sampleInt: (id) i;
- (NSNumber *) sampleInteger;
- (void) sampleInteger: (NSNumber *) i;
- (NSNumber *) sampleFloat;
- (void) sampleFloat: (NSNumber *) fl;
- (NSNumber *) sampleBoolean;
- (NSNumber *) pi;
- (NSString *) ch;
- (void) ch: (NSString *)s;
@end

@implementation TestFixture
- (NSString *) name {
    return name;
}

- (void) name: (NSString *) s {
    name = s;
}

- (void) ch: (NSString *)s {
    ch = [s characterAtIndex: 0];
}

- (void) sampleInt: (id) i {
    sampleInt = [i intValue];
}
- (int) sampleInt {
    return sampleInt;
}
- (NSNumber *) sampleInteger {
    return sampleInteger;
}
- (void) sampleInteger: (NSNumber *) i {
    sampleInteger = i;
}
- (void) sampleFloat: (NSNumber *) fl {
    sampleFloat = fl;
}
- (NSNumber *) sampleFloat {
    return sampleFloat;
}

- (void) sampleBoolean: (NSNumber *) b {
    sampleBoolean = [b boolValue];
}

- (NSNumber *) sampleBoolean {
    return [NSNumber numberWithBool: sampleBoolean];
}

- (NSNumber *) pi {
    return [NSNumber numberWithFloat: 3.14159862];
}
- (NSString *) ch {
    return [NSString stringWithCharacters: &ch length: 1];
}
@end

@implementation FrameworkTest

- (void) testParsing {
    NSArray *tags = [NSArray arrayWithObject: @"table"];

    Parse *p = [Parse initWithString: @"leader<Table foo=2>body</table>trailer" tags: tags];

    [self assert: [p leader] equals: @"leader"];
    [self assert: [p tag] equals: @"<Table foo=2>"];
    [self assert: [p body] equals: @"body"];
    [self assert: [p trailer] equals: @"trailer"];
}

- (void) testRecursing {
    NSArray *tags = [NSArray arrayWithObjects: @"table", @"tr", @"td", nil];
    Parse *p = [Parse initWithString: @"leader<table><TR><Td>body</tD></TR></table>trailer" tags: tags];
    [self assert: [p body] equals: nil];
    [self assert: [[p parts] body] equals: nil];
    [self assert: [[[p parts] parts] body] equals: @"body"];
}

- (void) testIterating {
    Parse *p = [Parse withString: @"leader<table><tr><td>one</td><td>two</td><td>three</td></tr></table>trailer"];
    [self assert: [[[p parts] parts] body] equals: @"one"];
    [self assert: [[[[p parts] parts] more] body] equals: @"two"];
    [self assert: [[[[[p parts] parts] more] more] body] equals: @"three"];
}

- (void) testIndexing {
    Parse *p = [Parse withString: @"leader<table><tr><td>one</td><td>two</td><td>three</td></tr><tr><td>four</td></tr></table>trailer"];
    [self assert: [[p at: 0 j: 0 k: 0] body] equals: @"one"];
    [self assert: [[p at: 0 j: 0 k: 1] body] equals: @"two"];
    [self assert: [[p at: 0 j: 0 k: 2] body] equals: @"three"];
    [self assert: [[p at: 0 j: 0 k: 3] body] equals: @"three"];
    [self assert: [[p at: 0 j: 0 k: 4] body] equals: @"three"];
    [self assert: [[p at: 0 j: 1 k: 0] body] equals: @"four"];
    [self assert: [[p at: 0 j: 1 k: 1] body] equals: @"four"];
    [self assert: [[p at: 0 j: 2 k: 0] body] equals: @"four"];
    [self assertInt: [p size] equals: 1];
    [self assertInt: [[p parts] size] equals: 2];
    [self assertInt: [[[p parts] parts] size] equals: 3];
    [self assert: [[p leaf] body] equals: @"one"];
    [self assert: [[[[p parts] last] leaf] body] equals: @"four"];
}

- (void) testParseException {
NS_DURING
    Parse *p = [Parse withString: @"leader<table><tr><th>one</th><th>two</th><th>three</th></tr><tr><td>four</td></tr></table>trailer"];
    [self assertFalse: p == nil];
    [self fail: @"expected exception not thrown"];
    NS_VOIDRETURN;
NS_HANDLER
    [self assert: [localException reason] equals: @"Can't find tag: td"];
NS_ENDHANDLER
}

- (void) testText {
    NSArray *tags = [NSArray arrayWithObjects: @"td", nil];
    Parse *p = [Parse initWithString: @"<td>a&lt;b</td>" tags: tags];
    [self assert: [p body] equals: @"a&lt;b"];
    [self assert: [p text] equals: @"a<b"];

    p = [Parse initWithString: @"<td>\ta&gt;b&nbsp;&amp;&nbsp;b>c &&&nbsp;</td>" tags: tags];
    [self assert: [p text] equals: @"a>b & b>c &&"];

    p = [Parse initWithString: @"<td>\ta&gt;b&nbsp;&amp;&nbsp;b>c &&nbsp;</td>" tags: tags];
    [self assert: [p text] equals: @"a>b & b>c &"];

    p = [Parse initWithString: @"<TD><P><FONT FACE=\"Arial\" SIZE=2>GroupTestFixture</FONT></TD>" tags: tags];
    [self assert: [p text] equals: @"GroupTestFixture"];
}

- (void) testUnescape {

    [self assert: [Parse unescape: @"a&lt;b"] equals: @"a<b"];
    [self assert: [Parse unescape: @"a&gt;b&nbsp;&amp;&nbsp;b>c &&"] equals: @"a>b & b>c &&"];
    [self assert: [Parse unescape: @"&amp;amp;&amp;amp;"] equals: @"&amp;&amp;"];
    [self assert: [Parse unescape: @"a&gt;b&nbsp;&amp;&nbsp;b>c &&"] equals: @"a>b & b>c &&"];
}

- (void) testUnformat {
    [self assert: [Parse unformat: @"<font size=+1>a</font>b"] equals: @"ab"];
    [self assert: [Parse unformat: @"a<font size=+1>b</font>"] equals: @"ab"];
    [self assert: [Parse unformat: @"a<b"] equals: @"a<b"];
}

- (void) testRuns {
    [self run: @"arithmetic" right: 37 wrong: 10 ignores: 0 exceptions: 2];
    [self run: @"CalculatorExample" right: 77 wrong: 7 ignores: 0 exceptions: 0];
    [self run: @"MusicExample" right: 95 wrong: 0 ignores: 0 exceptions: 0];
}

- (void) run: (NSString *) file right: (int) right wrong: (int) wrong ignores: (int) ignores exceptions: (int) exceptions {

    NSFileHandle *inHandle = [NSFileHandle fileHandleForReadingAtPath:
        [self directory: @"../Documents/" file: file suffix: @".html"]];

    NSFileManager *fm = [NSFileManager defaultManager];

    NSString *outFileName = [self directory: @"../Reports/" file: file suffix: @".html"];

    [fm createFileAtPath: outFileName contents: nil attributes: nil];

    NSFileHandle *outHandle = [NSFileHandle fileHandleForWritingAtPath: outFileName];

    NSData *data = [inHandle readDataToEndOfFile];

    NSString *content = [[NSString alloc] initWithData: data encoding: NSASCIIStringEncoding];

    Parse *p = [Parse withString: content];
    [content release];

    Fixture *fixture = [[Fixture alloc] init];
    [fixture autorelease];
    [fixture doTables: p];

    [p print: outHandle];
    [inHandle closeFile];
    [outHandle closeFile];

    [self assertInt: fixture->counts->right equals: right message: [file stringByAppendingString: @" right"]];
    [self assertInt: fixture->counts->wrong equals: wrong message: [file stringByAppendingString: @" wrong"]];
    
}

- (NSString *) directory: (NSString *) directory file: (NSString *) file suffix: (NSString *) suffix {
    return [[directory stringByAppendingString: file] stringByAppendingString: suffix];
}

//    protected void run(String file, int right, int wrong, int ignores, int exceptions) throws 
- (void) testTypeAdapter {
    TestFixture *f = [[TestFixture alloc] init];
    // don't know when I'll really be done with it
    [f autorelease];
    TypeAdapter *a = [TypeAdapter on: f field: @"sampleInt"];
    [a set: [a parse: @"123456"]];
    [self assertInt: [f sampleInt] equals: 123456];

    [self assert: [a parse: @"-234567"] equals: [NSNumber numberWithInt: [@"-234567" intValue]]];

    a = [TypeAdapter on: f field: @"sampleInteger"];
    [a set: [a parse: @"54321"]];
    [self assert: [f sampleInteger] equals: [NSNumber numberWithInt: [@"54321" intValue]]];

    a = [TypeAdapter on: f method: @"pi"];
    NSNumber *result = (NSNumber *)[a invokeMethod];
    [self assertFloat: [result floatValue] equals: 3.14159 precision: 0.00001];
    [self assert: [a invokeMethod] equals: [NSNumber numberWithFloat: 3.14159862]];

    a = [TypeAdapter on: f field: @"ch"];
    [a set: [a parse: @"abc"]];
    [self assert: [f ch] equals: @"a"];

    a = [TypeAdapter on: f field: @"name"];
    [a set: [a parse: @"xyzzy"]];
    [self assert: [f name] equals: @"xyzzy"];

    a = [TypeAdapter on: f field: @"sampleFloat"];
    [a set: [a parse: @"6.02e23"]];
    [self assertFloat: [[f sampleFloat] floatValue] equals: 6.02e23 precision: 1e17];

    a = [TypeAdapter on: f field: @"sampleBoolean"];
    [a set: [a parse: @"false"]];
    [self assertFalse: [[f sampleBoolean] boolValue] message: @"boolean failure"];
    
    /*
    a = TypeAdapter.on(f, f.getClass().getField("sampleArray"));
    a.set(a.parse("1,2,3"));
    assertEquals(1, f.sampleArray[0]);
    assertEquals(2, f.sampleArray[1]);
    assertEquals(3, f.sampleArray[2]);
    assertEquals("1, 2, 3", a.toString(f.sampleArray));
    assertTrue(a.equals(new int [] {1,2,3}, f.sampleArray));
    a = TypeAdapter.on(f, f.getClass().getField("sampleDate"));
    Date date = new Date(49,4,26);
    a.set(a.parse(DateFormat.getDateInstance().format(date)));
    assertEquals(date, f.sampleDate);
    a = TypeAdapter.on(f, f.getClass().getField("sampleByte"));
    a.set(a.parse("123"));
    assertEquals(123, f.sampleByte);
    a = TypeAdapter.on(f, f.getClass().getField("sampleShort"));
    a.set(a.parse("12345"));
    assertEquals(12345, f.sampleShort);
*/    
}

@end

File: dotnet/Source/fit/FrameworkTest.cs
// Copyright (c) 2002, 2003 Cunningham & Cunningham, Inc.
// Released under the terms of the GNU General Public License version 2 or later.

using System;
using NUnit.Framework;
using System.Reflection;
using System.IO;

namespace fit {
    public class FrameworkTest : TestCase {
        public FrameworkTest(string name) : base(name) {
        }

        public void TestParsing() {
            Parse p = new Parse("leader<Table foo=2>body</table>trailer", new string[] {"table"});
            AssertEquals("leader", p.leader);
            AssertEquals("<Table foo=2>", p.tag);
            AssertEquals("body", p.body);
            AssertEquals("trailer", p.trailer);
        }

        public void testRecursing() {
            Parse p = new Parse("leader<table><TR><Td>body</tD></TR></table>trailer");
            AssertEquals(null, p.body);
            AssertEquals(null, p.parts.body);
            AssertEquals("body", p.parts.parts.body);
        }

        public void testIterating() {
            Parse p = new Parse("leader<table><tr><td>one</td><td>two</td><td>three</td></tr></table>trailer");
            AssertEquals("one", p.parts.parts.body);
            AssertEquals("two", p.parts.parts.more.body);
            AssertEquals("three", p.parts.parts.more.more.body);
        }

        public void testIndexing() {
            Parse p = new Parse("leader<table><tr><td>one</td><td>two</td><td>three</td></tr><tr><td>four</td></tr></table>trailer");
            AssertEquals("one", p.at(0,0,0).body);
            AssertEquals("two", p.at(0,0,1).body);
            AssertEquals("three", p.at(0,0,2).body);
            AssertEquals("three", p.at(0,0,3).body);
            AssertEquals("three", p.at(0,0,4).body);
            AssertEquals("four", p.at(0,1,0).body);
            AssertEquals("four", p.at(0,1,1).body);
            AssertEquals("four", p.at(0,2,0).body);
            AssertEquals(1, p.size());
            AssertEquals(2, p.parts.size());
            AssertEquals(3, p.parts.parts.size());
            AssertEquals("one", p.leaf().body);
            AssertEquals("four", p.parts.last().leaf().body);
        }

        public void testParseException() {
            try {
                Parse p = new Parse("leader<table><tr><th>one</th><th>two</th><th>three</th></tr><tr><td>four</td></tr></table>trailer");
            } 
            catch (ApplicationException e) {
                //				AssertEquals(17, e.getErrorOffset());
                AssertEquals("Can't find tag: td", e.Message);
                return;
            }
            Fail("exptected exception not thrown");
        }

        public void testText() {
            string[] tags ={"td"};
            Parse p = new Parse("<td>a&lt;b</td>", tags);
            AssertEquals("a&lt;b", p.body);
            AssertEquals("a<b", p.text());
            p = new Parse("<td>\ta&gt;b&nbsp;&amp;&nbsp;b>c &&&nbsp;</td>", tags);
            AssertEquals("a>b & b>c &&", p.text());
            p = new Parse("<td>\ta&gt;b&nbsp;&amp;&nbsp;b>c &&nbsp;</td>", tags);
            AssertEquals("a>b & b>c &", p.text());
            p = new Parse("<TD><P><FONT FACE=\"Arial\" SIZE=2>GroupTestFixture</FONT></TD>", tags);
            AssertEquals("GroupTestFixture",p.text());
        }

        public void testUnescape () {
            AssertEquals("a<b", Parse.unescape("a&lt;b"));
            AssertEquals("a>b & b>c &&", Parse.unescape("a&gt;b&nbsp;&amp;&nbsp;b>c &&"));
            AssertEquals("&amp;&amp;", Parse.unescape("&amp;amp;&amp;amp;"));
            AssertEquals("a>b & b>c &&", Parse.unescape("a&gt;b&nbsp;&amp;&nbsp;b>c &&"));
        }

        public void testUnformat () {
            AssertEquals("ab",Parse.unformat("<font size=+1>a</font>b"));
            AssertEquals("ab",Parse.unformat("a<font size=+1>b</font>"));
            AssertEquals("a<b",Parse.unformat("a<b"));
        }

        public void TestTypeAdapter() {
            TypeAdapterTarget target = new TypeAdapterTarget();
            TypeAdapter adapter = TypeAdapter.on(target, typeof(int));
            object parsedInt = adapter.parse("123");
            AssertEquals("int", 123, (int)parsedInt);
            Assert("int should be equal", adapter.equals(parsedInt, 123));

            adapter = TypeAdapter.on(target, typeof(float));
            AssertEquals("float", 12.3, (float)adapter.parse("12.3"), .00001);

            adapter = TypeAdapter.on(target, typeof(long));
            AssertEquals("long", 123L, (long)adapter.parse("123"));

            adapter = TypeAdapter.on(target, typeof(string));
            AssertEquals("string", "123", (string)adapter.parse("123"));

            adapter = TypeAdapter.on(target, typeof(string[]));
            AssertArraysEqual("string[]", new object[] {"1", "2", "3"}, (object[])adapter.parse("1,2,3"));

            adapter = TypeAdapter.on(target, typeof(int[]));
            AssertArraysEqual("int[]", new object[] {1, 2, 3}, (object[])adapter.parse("1,2,3"));

            MethodInfo twoPi = typeof(TypeAdapterTarget).GetMethod("twoPi");
            adapter = TypeAdapter.on(target, twoPi);
            AssertEquals("twoPi", 2 * Math.PI, adapter.get());

            FieldInfo pi = target.GetType().GetField("pi");
            adapter = TypeAdapter.on(target, pi);
            adapter.set(3);
            AssertEquals("pi", 3, target.pi, .00001);
            AssertEquals("adapted pi", 3, (double)adapter.get(), .00001);
        }

        private void AssertArraysEqual(string message, object[] expected, object[] actual) {
            AssertEquals(message, Flatten(expected), Flatten(actual));
        }

        private string Flatten(object[] array) {
            string result = "";
            string comma = "";
            foreach (object o in array) {
                result += comma + o;
                comma = ",";
            }
            return result;
        }

        private class TypeAdapterTarget : Fixture {
            public double pi = Math.PI;

            public double twoPi() {
                return 2 * pi;
            }
        }
		
        public void testTypeAdapter () {
            TestFixture f = new TestFixture ();
            TypeAdapter a = TypeAdapter.on(f, f.GetType().GetField("sampleInt"));
            a.set(a.parse("123456"));
            AssertEquals(123456, f.sampleInt);
            AssertEquals("-234567", a.parse("-234567").ToString());
            a = TypeAdapter.on(f, f.GetType().GetMethod("pi", new Type[] {}));
            AssertEquals(3.14159, (double)a.get(), 0.00001);
            a = TypeAdapter.on(f, f.GetType().GetField("ch"));
            a.set(a.parse("a"));
            AssertEquals('a', f.ch);
            a = TypeAdapter.on(f, f.GetType().GetField("name"));
            a.set(a.parse("xyzzy"));
            AssertEquals("xyzzy", f.name);
            a = TypeAdapter.on(f, f.GetType().GetField("sampleFloat"));
            a.set(a.parse("6.02e23"));
            AssertEquals(6.02e23, f.sampleFloat, 1e17);
            //			a = TypeAdapter.on(f, f.GetType().GetField("sampleArray"));
            //			a.set(a.parse("1,2,3"));
            //			AssertEquals(1, f.sampleArray[0]);
            //			AssertEquals(2, f.sampleArray[1]);
            //			AssertEquals(3, f.sampleArray[2]);
            //			AssertEquals("1,2,3", f.sampleArray.ToString());
            //			AssertEquals(new int[] {1, 2, 3}, f.sampleArray);
            //			a = TypeAdapter.on(f, f.GetType().GetField("sampleDate"));
            //			DateTime date = new DateTime(1949,4,26);
            //			a.set(a.parse(DateFormat.getDateInstance().format(date)));
            //			AssertEquals(date, f.sampleDate);
            a = TypeAdapter.on(f, f.GetType().GetField("sampleByte"));
            a.set(a.parse("123"));
            AssertEquals((byte)123, f.sampleByte);
            a = TypeAdapter.on(f, f.GetType().GetField("sampleShort"));
            a.set(a.parse("12345"));
            AssertEquals((short)12345, f.sampleShort);
        }

        class TestFixture : ColumnFixture {
		  // used in testTypeAdapter
            public byte sampleByte = 0;
            public short sampleShort =0;
            public int sampleInt = 0;
            public float sampleFloat = 0;
            public double pi () {return 3.14159862;}
            public char ch = '\0';
            public string name = null;
            public int[] sampleArray = null;
            public DateTime sampleDate = DateTime.Now;
        }
        
        public void testScientificDouble() {
            double pi = 3.141592865;
            AssertEquals(ScientificDouble.valueOf("3.14"), pi);
            AssertEquals(ScientificDouble.valueOf("3.142"), pi);
            AssertEquals(ScientificDouble.valueOf("3.1416"), pi);
            AssertEquals(ScientificDouble.valueOf("3.14159"), pi);
            AssertEquals(ScientificDouble.valueOf("3.141592865"), pi);
            Assert(!ScientificDouble.valueOf("3.140").Equals(pi));
            Assert(!ScientificDouble.valueOf("3.144").Equals(pi));
            Assert(!ScientificDouble.valueOf("3.1414").Equals(pi));
            Assert(!ScientificDouble.valueOf("3.141592863").Equals(pi));
            AssertEquals(ScientificDouble.valueOf("6.02e23"), 6.02e23d);
            AssertEquals(ScientificDouble.valueOf("6.02E23"), 6.024E23d);
            AssertEquals(ScientificDouble.valueOf("6.02e23"), 6.016e23d);
            Assert(!ScientificDouble.valueOf("6.02e23").Equals(6.026e23d));
            Assert(!ScientificDouble.valueOf("6.02e23").Equals(6.014e23d));
            AssertEquals(ScientificDouble.valueOf("3.14"), ScientificDouble.valueOf("3.14"));
        }

        public void testEscape() {
            String junk = "!@#$%^*()_-+={}|[]\\:\";',./?`";
            AssertEquals(junk, Fixture.escape(junk));
            AssertEquals("", Fixture.escape(""));
            AssertEquals("&lt;", Fixture.escape("<"));
            AssertEquals("&lt;&lt;", Fixture.escape("<<"));
            AssertEquals("x&lt;", Fixture.escape("x<"));
            AssertEquals("&amp;", Fixture.escape("&"));
            AssertEquals("&lt;&amp;&lt;", Fixture.escape("<&<"));
            AssertEquals("&amp;&lt;&amp;", Fixture.escape("&<&"));
            AssertEquals("a &lt; b &amp;&amp; c &lt; d", Fixture.escape("a < b && c < d"));
        }

        public void testRuns() {
            run("arithmetic", 37, 10, 0, 2);
            run("CalculatorExample", 75, 9, 0, 0);
            run("MusicExample", 95, 0, 0, 0);
        }

        protected void run(string file, int right, int wrong, int ignores, int exceptions) {
            string input = read("../Documents/"+file+".html");
            Fixture fixture = new Fixture();
            Parse tables;
            if (input.IndexOf("<wiki>") >= 0) {
                tables = new Parse(input, new String[]{"wiki", "table", "tr", "td"});
                fixture.doTables(tables.parts);
            } else {
                tables = new Parse(input, new String[]{"table", "tr", "td"});
                fixture.doTables(tables);
            }
            StringWriter output = new StringWriter();
            tables.print(output);
            output.Close();

            AssertEquals(file+" right", right, fixture.counts.right);
            AssertEquals(file+" wrong", wrong, fixture.counts.wrong);
            AssertEquals(file+" ignores", ignores, fixture.counts.ignores);
            AssertEquals(file+" exceptions", exceptions, fixture.counts.exceptions);
        }

        protected String read(string filename) {
            StreamReader input = new StreamReader(filename);
            string result = input.ReadToEnd();
            input.Close();
            return result;
        }
    }
}